Migrate addon storyshots to TS

gaetanmaisse opened the pull request:
## What I did

 - Migrate `@storybook/addon-storyshots` and `@storybook/addon-storyshots-puppeteer` to TypeScript
 - Add types and classes to have a simpler DX when working on this addon
 - Refactor import/export to use named exports instead of default ones

Tracking Issue: https://github.com/storybooks/storybook/issues/5030

gaetanmaisse committed:
chore: emit TS declaration file during `addon-storyshots` build

gaetanmaisse committed:
refactor: move files of base and api part of `addon-storyshots` to TS

gaetanmaisse committed:
refactor: create StoryshotsOptions to export options available for addon-storyshots

Also improve a bit the typings

gaetanmaisse committed:
refactor: move addon-storyshots framework base file to TS and types them

gaetanmaisse committed:
refactor: move angular loader in addon-storyshots framework part to TS

gaetanmaisse committed:
refactor: move html loader of addon-storyshots framework part to TS

gaetanmaisse committed:
refactor: move preact loader of addon-storyshots framework part to TS

gaetanmaisse committed:
feat: make frameworkerLoader load both JS and TS loaders as they will be migrated to TS

gaetanmaisse committed:
refactor: move react loader of addon-storyshots framework part to TS

gaetanmaisse committed:
refactor: move riot loader of addon-storyshots framework part to TS

gaetanmaisse committed:
refactor: move react native loader of addon-storyshots framework part to TS

gaetanmaisse committed:
refactor: move svelte loader of addon-storyshots framework part to TS

gaetanmaisse committed:
refactor: move vue loader of addon-storyshots framework part to TS

gaetanmaisse committed:
refactor: extract SupportedFramework and use it to improve typings of storyshots options

gaetanmaisse committed:
refactor: used names export instead of default one for Stories2SnapsConverter

gaetanmaisse committed:
refactor: rename 'rn' folder to 'react-native'

Make react native folder name and so loader name consistent with others and also with storyshot "framework" option (that support 'react-native' string)

gaetanmaisse committed:
refactor: use `reduce` instead of `flatMap` because it is not available in Node 8

gaetanmaisse committed:
refactor: improve typings of some base classes of storyshots-core

gaetanmaisse commented:
@shilman @ndelangen should all stories from [storyshots-core](https://github.com/storybookjs/storybook/tree/741e69584eeb9a63c2e7a0c9e300a4d9537ee959/addons/storyshots/storyshots-core/stories) be migrated to TS? kept to JS? Duplicated to TS? 
I also read a bit about CSF format, maybe they should be converted to it?

gaetanmaisse committed:
chore: add tsconfig in `storyshots-puppeteer`

gaetanmaisse committed:
refactor: move all files of addon-storyshots-puppeteer to TS

gaetanmaisse committed:
refactor: improve typings of addon-storyshots-puppeteer

gaetanmaisse committed:
doc: fix `storyshots-puppeteer` README, so of the examples were not aligned with puppeteer attributes anymore

gaetanmaisse commented on addons/storyshots/storyshots-core/src/api/index.ts:
```diff
@@ -3,23 +3,27 @@ import addons, { mockChannel } from '@storybook/addons';
 import ensureOptionsDefaults from './ensureOptionsDefaults';
 import snapshotsTests from './snapshotsTestsTemplate';
 import integrityTest from './integrityTestTemplate';
+// @ts-ignore
```
👀

gaetanmaisse replied:
Resolved

gaetanmaisse commented on addons/storyshots/storyshots-core/src/api/index.ts:
```diff
@@ -3,23 +3,27 @@ import addons, { mockChannel } from '@storybook/addons';
 global.STORYBOOK_REACT_CLASSES = global.STORYBOOK_REACT_CLASSES || {};
 
-const methods = ['beforeAll', 'beforeEach', 'afterEach', 'afterAll'];
+type TestMethod = 'beforeAll' | 'beforeEach' | 'afterEach' | 'afterAll';
```
Useful?

gaetanmaisse commented on addons/storyshots/storyshots-core/src/api/index.ts:
```diff
@@ -3,23 +3,27 @@ import addons, { mockChannel } from '@storybook/addons';
+const methods: TestMethod[] = ['beforeAll', 'beforeEach', 'afterEach', 'afterAll'];
 
-function callTestMethodGlobals(testMethod) {
+function callTestMethodGlobals(testMethod: any) {
```
Need to investigate more, but it looks hard to type

gaetanmaisse commented on addons/storyshots/storyshots-core/src/api/index.ts:
```diff
@@ -45,11 +49,11 @@ function testStorySnapshots(options = {}) {
-    .filter(({ name }) => (storyNameRegex ? name.match(storyNameRegex) : true))
-    .filter(({ kind }) => (storyKindRegex ? kind.match(storyKindRegex) : true))
-    .reduce((acc, item) => {
+    .filter(({ name }: any) => (storyNameRegex ? name.match(storyNameRegex) : true))
```
`string`?

gaetanmaisse commented on addons/storyshots/storyshots-core/src/api/index.ts:
```diff
@@ -45,11 +49,11 @@ function testStorySnapshots(options = {}) {
-    .filter(({ kind }) => (storyKindRegex ? kind.match(storyKindRegex) : true))
-    .reduce((acc, item) => {
+    .filter(({ name }: any) => (storyNameRegex ? name.match(storyNameRegex) : true))
+    .filter(({ kind }: any) => (storyKindRegex ? kind.match(storyKindRegex) : true))
```
`string`?

gaetanmaisse commented on addons/storyshots/storyshots-core/src/frameworks/react-native/loader.ts:
```diff
@@ -21,20 +23,22 @@ function configure(options, storybook) {
 }
 
-export default {
+const rnLoader: Loader = {
```
To rename

gaetanmaisse commented on addons/storyshots/storyshots-puppeteer/src/imageSnapshot.ts:
```diff
@@ -34,10 +35,10 @@ export const imageSnapshot = (customConfig = {}) => {
-  const testFn = async ({ context }) => {
+  const testFn = async ({ context }: any) => {
     const { kind, framework, name } = context;
     if (framework === 'rn') {
```
to rename according to https://github.com/storybookjs/storybook/pull/7674/files#diff-0b05142d970443de5c41e81ad14df803R7?

gaetanmaisse replied:
Moreover it looks like it match https://github.com/storybookjs/storybook/pull/7674/files#diff-d72af9277db9c8a6cb1b3595750de8c4R9 which is not currently true

gaetanmaisse commented on addons/storyshots/storyshots-core/src/api/StoryshotsOptions.ts:
```diff
@@ -0,0 +1,23 @@
+  storyKindRegex?: RegExp | string;
+  storyNameRegex?: RegExp | string;
+  framework?: SupportedFramework;
+  test?: (story: any, context: any, renderTree: RenderTree, options?: any) => any;
```
@ndelangen @kroeder are there already existing types somewhere for `story` and `context`?

ndelangen replied:
Yes, in lib/addons

gaetanmaisse commented on addons/storyshots/storyshots-core/package.json:
```diff
@@ -26,6 +26,9 @@
     "@storybook/addons": "5.2.0-beta.22",
+    "@types/glob": "^7.1.1",
+    "@types/jest": "^24.0.16",
+    "@types/jest-specific-snapshot": "^0.5.3",
```
@emilio-martinez @kroeder @ndelangen what's the best practice about `@types` in SB codebase? I added them in `dependencies` because some of the types I exported in this addon are based on `jest` (and other libs) ones. So if a SB user want to use SB type it should install `@types/jest` too... 🤷‍♂

gaetanmaisse commented on addons/storyshots/storyshots-puppeteer/package.json:
```diff
@@ -29,10 +29,14 @@
   "peerDependencies": {
     "@storybook/addon-storyshots": "5.2.0-beta.13",
+    "@types/puppeteer": "^1.19.0",
+    "@types/jest-image-snapshot": "^2.8.0",
```
Same question as https://github.com/storybookjs/storybook/pull/7674/files#r312234521

gaetanmaisse commented on addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts:
```diff
@@ -42,7 +44,9 @@ function load(options) {
 }
 
-export default {
+const angularLoader: Loader = {
```
It's the way I found to have type-check according to `Loader` class, if anyone has a better proposition?

shilman commented:
@gaetanmaisse @ndelangen Let's merge this after 5.2 is released please 🙏

ndelangen commented:
hey @gaetanmaisse I merged in next, do you think you'd be able to finish this PR?

gaetanmaisse commented:
@ndelangen I just come back from holidays I'm on it 😉

gaetanmaisse committed:
refactor: improve storyshot-core typings

gaetanmaisse committed:
fix: fix `react-native` string used in framework check of imageSnapshot

Framework name should match the one used in framework loader (in this case React Native loader: https://github.com/storybookjs/storybook/pull/7674/files#diff-d72af9277db9c8a6cb1b3595750de8c4R9)

gaetanmaisse committed:
feat: use `puppeteer-core` instead of `puppeteer`

`puppeteer-core` package is a version of `puppeteer` that doesn't download Chromium by default.
It looks like chromium is not directly used here as it was only a `peerDependencies` or an `optionalDependencies`.
So I moved to `puppeteer-core` to be able to compile TS sources of this addon without downloading Chromium at every `yarn install`

gaetanmaisse commented:
A lot of types are still missing but I think I reach a good state to merge this PR and continue my work on this after it will be merged.

During reviews, a focus on my last commit would be great because I think this part is not tested during PR automatic tests (but I'm maybe wrong?) and I may have incorrectly understood how this puppeteer related addon works.

ndelangen commented:
What's the difference between puppeteer & puppeteer-core?

ndelangen commented:
So awesome you got this green 🎉

gaetanmaisse commented:
@ndelangen 
> What's the difference between puppeteer & puppeteer-core?

See commit message of https://github.com/storybookjs/storybook/pull/7674/commits/749c3944d2cd7ef8dfb8fd6837706e6e6d244e0a : 
>`puppeteer-core` package is a version of `puppeteer` that doesn't download Chromium by default.
It looks like chromium is not directly used here as it was only a `peerDependencies` or an `optionalDependencies`.
So I moved to `puppeteer-core` to be able to compile TS sources of this addon without downloading Chromium at every `yarn install`

SimenB commented:
https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteer-vs-puppeteer-core

ndelangen commented:
Awesome, merge when you think it's ready!

shilman reviewed (commented):
More `any` types than I'd expect, but I'm assuming there's good reason for it?

gaetanmaisse commented on addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts:
```diff
@@ -18,13 +20,13 @@ function setupAngularJestPreset() {
 }
 
-function test(options) {
+function test(options: any): boolean {
```
Good catch! I fixed it.

gaetanmaisse commented:
@shilman 
> More any types than I'd expect, but I'm assuming there's good reason for it?

A lot of them are related to `story` or `context` function arguments, I tried to use some types defined in `lib/addons` but at the end, I have some conflicting/incompatible types 😞 So what I have in mind is merging this PR to share the migration to everybody. Then continue to work to improve typings of `storyshot` addons and make a 2nd PR when it will be ready.
